---
title: "Testing scanorama"
author: "Friederike Duendar"
date: "5/16/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=FALSE)
```

The scanorama code can be found on github: [https://github.com/brianhie/scanorama](https://github.com/brianhie/scanorama).
The article was published in May 2019: [Hie, Bryson, Berger](https://www.nature.com/articles/s41587-019-0113-3).

>Our approach is analogous to computer vision algorithms for panorama stitching that identify images with overlapping content and merge these into a larger panorama.
> Likewise, Scanorama automatically identifies scRNA-seq datasets containing cells with similar transcriptional profiles and can leverage those matches for batch correction and integration (Fig. 1b), without also merging datasets that do not overlap.

* searches nearest neighbors using dimensionality reduction and approximate nearest-neighbors 
* mutually linked cells form matches that can be leveraged to correct for batch effects
* scRNA-seq "panorama" = datasets forming connected components on the basis of the kNN-matches
* insensitive to order and less vulnerable to overcorrection because it finds matches between all pairs of datasets.

```{r  eval=FALSE}
# /scratchLocal/frd2007/software/anaconda3/bin/R
all <- readRDS("sce_2018-08-02_filteredCellsGenes_countsOnly.rds")
```

```{r prep_data, eval=FALSE}
## list of cells-by-genes matrices
datasets <- lapply(c("WT","DB","HFD"), function(x){
  out <- all[, grepl(x, colnames(all))]
  out <- t(out)
  return(out)
  })

## list of gene names
#names(datasets) <- c("WT","DB","HFD") ## named lists break the reticulate translation!
genes_list <- lapply(datasets, colnames)

```

Scanorama was written in Python. I've installed it with conda/pip within the conda environment "scrna".

```{r engine="bash", eval=FALSE}
conda activate scrna # make sure to switch to that environment
conda install pip
/scratchLocal/frd2007/software/anaconda3/envs/scrna/bin/pip install scanorama
```

The integration with `reticulate` shown here was based on the info gleaned from [this](https://github.com/brianhie/scanorama/blob/master/bin/R/scanorama.R) file
by the scanorama author.

```{r}
library(reticulate)
use_condaenv("scrna")
scanorama <- import('scanorama')
```

### Integrating the data

```{r}
integrated.data <- scanorama$integrate(datasets, genes_list)
```

```
> str(integrated.data)
List of 2
 $ :List of 3
  ..$ : num [1:4337, 1:100] -0.0509 -0.0727 -0.0562 -0.1066 -0.2198 ...
  ..$ : num [1:5839, 1:100] 0.2667 0.2695 -0.0851 0.2072 0.1265 ...
  ..$ : num [1:2835, 1:100] -0.0846 -0.1927 -0.424 -0.4013 -0.2506 ...
 $ : chr [1:17281(1d)] "ENSMUSG00000000001" "ENSMUSG00000000028" "ENSMUSG00000000031" "ENSMUSG00000000037" ...
```

From the `help(scanorama.integrate)` page:

> Returns a two-tuple containing a list of `numpy.ndarray` with **integrated low dimensional embeddings** and a single list of genes containing the intersection of inputted genes.

Parameters of `scanorama.integrate()`:

| Option | Default | Meaning |
|--------|---------|---------|
| `datasets_full` | |  Data sets to integrate and correct. | 
| `genes_list` | |  List of genes for each data set. |
| `batch_size` | 5000 |   The batch size used in the alignment vector computation. Useful when correcting very large (>100k samples) data sets. Set to large value that runs within available memory. |
| `dimred` | 100|  Dimensionality of integrated embedding. |
| `approx` | True | Use approximate nearest neighbors, greatly speeds up matching runtime.|
| `sigma` | 15 |  Correction smoothing parameter on Gaussian kernel. |
| `alpha` | 0.1 | Alignment score minimum cutoff.|
| `knn` |20 | Number of nearest neighbors to use for matching. |
| `geosketch` | False  | |
| `geosketch_max` | 20000 | |
| `n_iter` | 1 | |
| `union` | False| | 
| `hvg` | None |  Use this number of top highly variable genes based on dispersion. |

    
    

### Batch correction

Obtain a matrix of batch-corrected values.

```{r}
corrected.data <- scanorama$correct(datasets, genes_list, return_dense=TRUE)
```

```
> str(corrected.data)
List of 2
 $ :List of 3
  ..$ : num [1:4337, 1:17281] 2.20e-04 1.92e-05 1.77e-05 1.64e-05 1.66e-05 ...
  ..$ : num [1:5839, 1:17281] 2.71e-04 6.28e-05 4.65e-05 6.58e-05 5.86e-05 ...
  ..$ : num [1:2835, 1:17281] 3.36e-05 3.58e-05 4.41e-05 5.40e-05 3.28e-04 ...
 $ : chr [1:17281(1d)] "ENSMUSG00000000001" "ENSMUSG00000000028" "ENSMUSG00000000031" "ENSMUSG00000000037" ...
```

From the `help(scanorama.correct)` page:


> By default (`return_dimred=False`), returns a two-tuple containing a
  **list of `scipy.sparse.csr_matrix` each with batch corrected values**,
  and a single list of genes containing the intersection of inputted
  genes.
    
> When `return_dimred=False`, returns a three-tuple containing a list
  of `numpy.ndarray` with integrated **low dimensional embeddings**, a list
  of `scipy.sparse.csr_matrix` each with **batch corrected values**, and a
  a single list of genes containing the intersection of inputted genes.


Parameters of `scanorama.correct()`:

| Option | Default | Meaning |
|--------|---------|---------|
| `datasets_full` | | Data sets to integrate and correct |
| `genes_list`    | | List of genes for each data set. |
| `return_dimred` | False| In addition to returning batch corrected matrices, also returns integrated low-dimesional embeddings |
| `batch_size` | 5000 | The batch size used in the alignment vector computation. Useful when correcting very large (>100k samples) data sets. Set to large value that runs within available memory. |
| `dimred` | 100 |  Dimensionality of integrated embedding. |
| `approx` | True |  Use approximate nearest neighbors, greatly speeds up matching runtime. |
| `sigma` | 15| Correction smoothing parameter on Gaussian kernel. |
| `alpha` | 0.1 |  Alignment score minimum cutoff. |
| `knn` | 20|  Number of nearest neighbors to use for matching. |
|`return_dense` | False | Return `numpy.ndarray` matrices instead of `scipy.sparse.csr_matrix`. |
| `hvg` | None | Use this number of top highly variable genes based on dispersion. |
| `union` | False |
| `geosketch` | False |
| `geosketch_max` | 20000 |


### Integration and batch correction

Obtain a matrix of batch-corrected values in addition to the low-dimensionality embeddings.

  
```{r}
integrated.corrected.data <- scanorama$correct(datasets, genes_list,
                                               return_dimred=TRUE, 
                                               return_dense=TRUE)

```

```
> str(integrated.corrected.data)
List of 3
 $ :List of 3
  ..$ : num [1:4337, 1:100] -0.0513 -0.0731 -0.0565 -0.107 -0.2204 ...
  ..$ : num [1:5839, 1:100] 0.2674 0.2699 -0.0853 0.2077 0.1268 ...
  ..$ : num [1:2835, 1:100] -0.0854 -0.1937 -0.4252 -0.4026 -0.2517 ...
 $ :List of 3
  ..$ : num [1:4337, 1:17281] 2.22e-04 2.13e-05 1.97e-05 1.86e-05 1.95e-05 ...
  ..$ : num [1:5839, 1:17281] 2.71e-04 6.26e-05 4.81e-05 6.52e-05 5.90e-05 ...
  ..$ : num [1:2835, 1:17281] 3.54e-05 3.83e-05 4.84e-05 6.08e-05 3.31e-04 ...
 $ : chr [1:17281(1d)] "ENSMUSG00000000001" "ENSMUSG00000000028" "ENSMUSG00000000031" "ENSMUSG00000000037" ...
```



